File: ruby/tictactoe/lib/tictactoe/models/game.rb
module Tictactoe
  module Models
    class Game
      VALID_SYMBOLS = %w(O X)
      attr_accessor :players, :board, :last_turn, :winner, :draw

      def initialize(board_size)
        self.board = Tictactoe::Models::Board.new(board_size)
        self.players = []
        self.draw = false
      end

      def add_player(name, symbol)
        check_if_more_than_two_players!
        check_invalid_symbol!(symbol)
        check_duplicate_symbol!(symbol)
        check_duplicate_player_name!(name)

        player = Player.new(name, symbol)
        players << player
        self.last_turn ||= player
      end

      def play(x, y)
        puts "place [#{last_turn.symbol}] for #{last_turn.name} at [#{x}],[#{y}]"
        board.place(last_turn.symbol, x, y)
        board.print if Tictactoe.print_matrix_on_each_move?
        check_winner?
        check_draw?
        switch_turn
      end

      private def check_winner?
        if board.has_winner?
          self.winner = last_turn
          puts "Winner: #{winner.name}"
        end
      end

      private def check_draw?
        if board.draw?
          self.draw = true
          puts "Game Draw"
        end
      end

      private def switch_turn
        self.last_turn = (last_turn == players[0] ? players[1] : players[0])
      end

      private def check_invalid_symbol!(symbol)
        raise Tictactoe::Errors::InvalidSymbolError unless VALID_SYMBOLS.include?(symbol.upcase)
      end

      private def check_duplicate_symbol!(symbol)
        return if players.empty?

        raise Tictactoe::Errors::DuplicateSymbolError if players[0].symbol == symbol
      end

      private def check_duplicate_player_name!(name)
        return if players.empty?

        raise Tictactoe::Errors::DuplicatePlayerNameError if players[0].name == name
      end

      private def check_if_more_than_two_players!
        raise Tictactoe::Errors::MoreThanTwoPlayersNotSupportedError if players.count == 2
      end
    end
  end
end

File: java/tictactoe/src/main/java/com/gagan93jtg/lld_learning/models/Game.java
package com.gagan93jtg.lld_learning.models;

import com.gagan93jtg.lld_learning.errors.*;
import com.gagan93jtg.lld_learning.services.Runner;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Game {
  public static final String[] VALID_SYMBOLS = {"O", "X"};
  List players = new ArrayList<Player>();
  Board board;
  Player lastTurn;
  String winner;
  boolean isDraw = false;

  public Game(int boardSize) {
    this.board = new Board((boardSize));
  }

  public void addPlayer(String name, String symbol) throws MoreThanTwoPlayersNotSupportedError, InvalidSymbolError, DuplicateSymbolError, PlayerInformationMissingError {
    checkIfMorethanTwoPlayers();
    checkIfSymbolInvalid(symbol);
    checkDuplicateSymbol(symbol);
    checkDuplicatePlayerName(name);

    Player player = new Player(name, symbol);
    if (players.isEmpty()) {
      lastTurn = player;
    }
    players.add(player);
  }

  public void play(int x, int y) throws InvalidCoordinatesError, BoardCellOccupiedError {
    board.place(lastTurn.symbol, x, y);
    if (Runner.printMatrix()) {
      board.print();
    }
    checkWinner();
    checkDraw();
    switchTurn();
  }

  public List<Player> getPlayers() {
    return players;
  }

  private void checkWinner() {
    if (board.hasWinner()) {
      this.winner = lastTurn.name;
      System.out.println("Winner: " + winner);
    }
  }

  private void checkDraw() {
    if (board.hasDraw()) {
      this.isDraw = true;
      System.out.println("Game Draw !");
    }
  }

  private void switchTurn() {
    this.lastTurn = (Player) (lastTurn == players.get(0) ? players.get(1) : players.get(0));
  }

  private void checkIfSymbolInvalid(String symbol) throws InvalidSymbolError {
    if (!Arrays.asList(VALID_SYMBOLS).contains(symbol)) {
      throw new InvalidSymbolError("The symbol: " + symbol + " is not valid.");
    }
  }

  private void checkDuplicateSymbol(String symbol) throws DuplicateSymbolError {
    if (players.isEmpty())
      return;

    Player firstPlayer = (Player)(players.get(0));
    if (firstPlayer.symbol == symbol) {
      throw new DuplicateSymbolError("Symbol already used by some other player");
    }
  }

  private void checkDuplicatePlayerName(String name) throws DuplicateSymbolError {
    if (players.isEmpty())
      return;

    Player firstPlayer = (Player)(players.get(0));
    if (firstPlayer.name == name) {
      throw new DuplicateSymbolError("Name already used by some other player");
    }
  }

  private void checkIfMorethanTwoPlayers() throws MoreThanTwoPlayersNotSupportedError {
    if (players.size() == 2) {
      throw new MoreThanTwoPlayersNotSupportedError("More than two players not supported");
    }
  }
}

